File: gallery/src/main/java/com/qkopy/gallery/ui/camera/DefaultCameraModule.java
/*
 * Created by @ajithvgiri on 4/12/18 8:11 PM
 * Copyright (c) 2018 . All rights reserved.
 * Last modified 4/12/18 7:05 PM
 */

package com.qkopy.gallery.ui.camera;

import android.content.Context;
import android.content.Intent;
import android.media.MediaScannerConnection;
import android.net.Uri;
import android.provider.MediaStore;
import androidx.core.content.FileProvider;
import com.qkopy.gallery.helper.ImageHelper;
import com.qkopy.gallery.model.Config;

import java.io.File;
import java.io.Serializable;
import java.util.Locale;

public class DefaultCameraModule implements CameraModule, Serializable {

    protected String imagePath;


    @Override
    public Intent getCameraIntent(Context context, Config config) {
        Intent intent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        File imageFile = new ImageHelper().createImageFile(config.getSavePath());
        if (imageFile != null) {
            Context appContext = context.getApplicationContext();
            String providerName = String.format(Locale.ENGLISH, "%s%s", appContext.getPackageName(), ".fileprovider");
            Uri uri = FileProvider.getUriForFile(appContext, providerName, imageFile);
            imagePath = imageFile.getAbsolutePath();
            intent.putExtra(MediaStore.EXTRA_OUTPUT, uri);
            ImageHelper.grantAppPermission(context, intent, uri);
            return intent;
        }
        return null;
    }

    @Override
    public void getImage(final Context context, Intent intent, final OnImageReadyListener imageReadyListener) {
        if (imageReadyListener == null) {
            throw new IllegalStateException("OnImageReadyListener must not be null");
        }

        if (imagePath == null) {
            imageReadyListener.onImageReady(null);
            return;
        }

        final Uri imageUri = Uri.parse(new File(imagePath).toString());
        if (imageUri != null) {
            MediaScannerConnection.scanFile(context.getApplicationContext(), new String[]{imageUri.getPath()}
                    , null
                    , new MediaScannerConnection.OnScanCompletedListener() {
                        @Override
                        public void onScanCompleted(String path, Uri uri) {
                            if (path != null) {
                                path = imagePath;
                            }
                            imageReadyListener.onImageReady(ImageHelper.singleListFromPath(path));
                            ImageHelper.revokeAppPermission(context, imageUri);
                        }
                    });
        }
    }
}

File: gallery/src/main/java/com/qkopy/gallery/ui/camera/DefaultCameraModule.kt
package com.qkopy.gallery.ui.camera

import android.content.Context
import android.content.Intent
import android.media.MediaScannerConnection
import android.net.Uri
import android.provider.MediaStore
import androidx.core.content.FileProvider
import com.qkopy.gallery.helper.ImageHelper
import com.qkopy.gallery.model.Config
import java.io.File
import java.io.Serializable
import java.util.*

class DefaultCameraModule : CameraModule, Serializable {
    protected var imagePath: String? = null


    override fun getCameraIntent(context: Context?, config: Config?): Intent? {
        val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        val imageFile: File = ImageHelper.createImageFile(context!!, config!!.savePath!!)!!
        if (imageFile != null) {
            val appContext = context!!.applicationContext
            val providerName = String.format(
                Locale.ENGLISH,
                "%s%s",
                appContext.packageName,
                ".fileprovider"
            )
            val uri =
                FileProvider.getUriForFile(appContext, providerName, imageFile)
            imagePath = imageFile.absolutePath
            intent.putExtra(MediaStore.EXTRA_OUTPUT, uri)
            ImageHelper.grantAppPermission(context, intent, uri)
            return intent
        }
        return null
    }

    override fun getImage(
        context: Context?,
        intent: Intent?,
        imageReadyListener: OnImageReadyListener?
    ) {
        checkNotNull(imageReadyListener) { "OnImageReadyListener must not be null" }
        if (imagePath == null) {
            imageReadyListener.onImageReady(null)
            return
        }
        val imageUri = Uri.parse(File(imagePath).toString())
        if (imageUri != null) {
            MediaScannerConnection.scanFile(
                context!!.applicationContext,
                arrayOf(imageUri.path),
                null
            ) { path, uri ->
                var path = path
                if (path != null) {
                    path = imagePath
                }
                imageReadyListener.onImageReady(ImageHelper.singleListFromPath(path))
                ImageHelper.revokeAppPermission(context, imageUri)
            }
        }
    }
}
